core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/util/ProgressListenerUtils.java

```diff
@@ -0,0 +1,102 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.util;
+
+import static software.amazon.awssdk.http.Header.CONTENT_LENGTH;
+
+import java.util.concurrent.atomic.AtomicLong;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.RequestOverrideConfiguration;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
+import software.amazon.awssdk.core.internal.http.RequestExecutionContext;
+import software.amazon.awssdk.core.internal.metrics.BytesReadTrackingInputStream;
+import software.amazon.awssdk.core.internal.metrics.BytesReadTrackingPublisher;
+import software.amazon.awssdk.core.internal.progress.listener.ProgressUpdater;
+import software.amazon.awssdk.http.AbortableInputStream;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
+import software.amazon.awssdk.utils.StringUtils;
+
+@SdkInternalApi
+public final class ProgressListenerUtils {
+
+    private ProgressListenerUtils() {
+    }
+
+    public static SdkHttpContentPublisher wrapRequestProviderWithByteTrackingIfProgressListenerAttached(
+        SdkHttpContentPublisher requestProvider, RequestExecutionContext context) {
+        ProgressUpdater progressUpdater = getProgressUpdaterIfAttached(context);
+        SdkHttpContentPublisher wrappedHttpContentPublisher = requestProvider;
+        if (progressUpdater != null) {
+            wrappedHttpContentPublisher = new BytesReadTrackingPublisher(requestProvider, new AtomicLong(0L),
+                                                                         new UploadProgressUpdaterInvocation(progressUpdater));
+        }
+        return wrappedHttpContentPublisher;
+    }
+
+    public static ContentStreamProvider wrapContentStreamProviderWithBytePublishTrackingIfProgressListenerAttached(
+        ContentStreamProvider contentStreamProvider, RequestExecutionContext context) {
+
+        ProgressUpdater progressUpdater = getProgressUpdaterIfAttached(context);
+        ContentStreamProvider wrappedContentStreamProvider = contentStreamProvider;
+        if (progressUpdater != null) {
+            wrappedContentStreamProvider =
+                ContentStreamProvider.fromInputStream(new BytesReadTrackingInputStream(
+                    AbortableInputStream.create(contentStreamProvider.newStream()),
+                    new AtomicLong(0L),
+                    new UploadProgressUpdaterInvocation(progressUpdater)));
```

Comment:
This is a bit hard to read, let's create variables.

Reply:
Looks like the existing implementation passes `RESPONSE_BYTES_READ` to  `BytesReadTrackingInputStream` and now we are passing 0L, is this expected?

Reply:
RESPONSE_BYTES_READ is used to account for "The running count of bytes in the response body that have been read by the client." Here, we are using it for an upload context but the same BytesReadTrackingInputStream is made use of. Hence, the 0.

Reply:
I think this will break existing code because we will later read `context.executionAttributes().getAttribute(SdkInternalExecutionAttribute.RESPONSE_BYTES_READ)`  to get the counter